### PR TITLE
Expand softmax for usage in FCNs

### PR DESCRIPTION
**Pull request by MartinThoma**
Related to https://github.com/fchollet/keras/issues/1169#issuecomment-220265092

### CODE REVIEW

File: keras/activations.py [hunk 1]
```diff
@@ -4,14 +4,37 @@
...
+    """
     ndim = K.ndim(x)
     if ndim == 2:
         return K.softmax(x)
     elif ndim == 3:
         e = K.exp(x - K.max(x, axis=-1, keepdims=True))
         s = K.sum(e, axis=-1, keepdims=True)
         return e / s
+    elif ndim == 4:
+        if 'channels_last':
```
Comment (joelthchao) [inline keras/activations.py]: `if data_format == 'channels_last':

File: keras/activations.py [hunk 2]
```diff
@@ -4,14 +4,37 @@
 from .utils.generic_utils import deserialize_keras_object
 
 
-def softmax(x):
+def softmax(x, data_format='channels_last'):
```
Comment (joelthchao) [inline keras/activations.py]: Maybe use `K.image_data_format()` to detect user preference?

File: keras/activations.py [hunk 3]
```diff
@@ -4,14 +4,37 @@
 from .utils.generic_utils import deserialize_keras_object
 
 
-def softmax(x):
+def softmax(x, data_format=K.image_data_format()):
+    """
```
Comment (joelthchao) [inline keras/activations.py]: Better way to do this is:
```python
def softmax(x, data_format=None):
    # ...
    elif ndim == 4:
        if not data_format:
            data_format = K.image_data_format()
        if data_format == 'channels_last':
            # ...
```
Both allow user to specify `data_format` and fallback to use `K.image_data_format()`

**Comment by fchollet** [comment]
Too many assumptions; this would break people who process timeseries data while using the default `channels_first`.

**Comment by joelthchao** [comment]
@fchollet Just to confirm, using TimeDistributed layer with this modified softmax will fail?

**Comment by fchollet** [comment]
I would recommend instead having an `axis` argument, defaulting to `-1`. For FCNs with `data_format="channels_first"`, you can then use `axis=1`.

File: keras/activations.py [hunk 4]
```diff
@@ -4,14 +4,24 @@
 from .utils.generic_utils import deserialize_keras_object
 
 
-def softmax(x):
+def softmax(x, axis=-1):
+    """
```
Comment (fchollet) [inline keras/activations.py]: Please add a one-line description (all docstrings must start with one) and a `# Returns` section.

File: keras/activations.py [hunk 5]
```diff
@@ -4,14 +4,24 @@
...
+        axis: integer, defines along which axis the normalization is applied
+    """
     ndim = K.ndim(x)
     if ndim == 2:
         return K.softmax(x)
     elif ndim == 3:
         e = K.exp(x - K.max(x, axis=-1, keepdims=True))
         s = K.sum(e, axis=-1, keepdims=True)
         return e / s
+    elif ndim == 4:
```
Comment (fchollet) [inline keras/activations.py]: You don't need this case. It's just either 2D -> `K.softmax`, > 2D -> use `axis`.

File: keras/activations.py [hunk 6]
```diff
@@ -4,13 +4,25 @@
...
         return K.softmax(x)
-    elif ndim == 3:
-        e = K.exp(x - K.max(x, axis=-1, keepdims=True))
-        s = K.sum(e, axis=-1, keepdims=True)
+    elif ndim > 2:
+        e = K.exp(x - K.max(x, axis=axis, keepdims=True))
+        s = K.sum(e, axis=axis, keepdims=True)
         return e / s
     else:
         raise ValueError('Cannot apply softmax to a tensor '
```
Comment (joelthchao) [inline keras/activations.py]: There is only one case, `ndim < 2`, to raise `ValueError` and error message should be fixed.

**Comment by fchollet** [comment]
LGTM, thanks

**Comment by tawantinsuyu** [comment]
Hi, 

How do we use this in the model? Do we use  Activation with keyword arguments activation="softmax", and axis=1? Or do we need to use Lambda layer?

Thanks
